src/server/pfs/server/driver.go

```diff
@@ -421,6 +424,57 @@ func (d *driver) deleteRepo(txnCtx *txncontext.TransactionContext, repo *pfs.Rep
 	return nil
 }
 
+func (d *driver) createProject(ctx context.Context, req *pfs.CreateProjectRequest) error {
+	if err := ancestry.ValidateName(req.Project.Name); err != nil {
+		return errors.Wrapf(err, "invalid project name")
+	}
+	if err := d.env.TxnEnv.WithWriteContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
+		projects := d.projects.ReadWrite(txnCtx.SqlTx)
+		if req.Update {
+			if err := projects.Get(pfsdb.ProjectKey(req.Project), &pfs.ProjectInfo{}); err != nil {
+				return errors.Wrapf(err, "update project")
+			}
+			return errors.EnsureStack(projects.Put(pfsdb.ProjectKey(req.Project), &pfs.ProjectInfo{
+				Project:     req.Project,
+				Description: req.Description,
+			}))
+		}
+		return errors.EnsureStack(projects.Create(pfsdb.ProjectKey(req.Project), &pfs.ProjectInfo{
+			Project:     req.Project,
+			Description: req.Description,
+		}))
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *driver) inspectProject(ctx context.Context, project *pfs.Project) (*pfs.ProjectInfo, error) {
+	pi := &pfs.ProjectInfo{}
+	if err := d.projects.ReadOnly(ctx).Get(pfsdb.ProjectKey(project), pi); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	return pi, nil
+}
+
+func (d *driver) listProject(ctx context.Context, cb func(*pfs.ProjectInfo) error) error {
+	projectInfo := &pfs.ProjectInfo{}
+	if err := d.projects.ReadOnly(ctx).List(projectInfo, col.DefaultOptions(), func(string) error {
+		return cb(proto.Clone(projectInfo).(*pfs.ProjectInfo))
```

Comment:
Is the clone necessary?

Reply:
Right now it is because we collect projects into a list during DeleteAll. List Repo works the same way. The client could also be responsible for copying but I think this is nice + consistent

Reply:
That should probably happen in DeleteAll then. It seems like the cloning should only happen where it is needed. Although I don't think the performance will be a big deal since the number of projects will probably never be high enough for that to matter, a separate use case for list project would incur that overhead even if it doesn't need it.